### ISSUE TITLE
SMALL_DELTA_CHI2 for different star templates

### ISSUE THREAD

**Issue by londumas** [desihub/redrock#47]
A non negligible number of stars (from eye scan) are flagged as ZWARN!=0 because another template of star fits nearly as well.
Since the star classification is a continuum, it is normal that for some stars we get two compatible templates.
We could set a condition stating that if the second best template is also a star then SMALL_DELTA_CHI2=0

For example target=3586551810094: 
```
znum spectype         z                 zerr       zwarn      chi2    
---- -------- ------------------ ----------------- ----- -------------
   0     STAR (A) -0.000116860579997  1.9342047008e-05     4 4007.82786385
   1     STAR (B) -0.000114349129022 1.95406225734e-05     0 4015.84137551
   2     STAR (F) -0.000120274130256 1.95726752171e-05     0   4311.382544
   4      QSO      1.27647633093 0.000268442671122     0 8235.95752663
   5      QSO      1.15285765716 0.000289393887284     0 8548.70757259
   6      QSO     0.737408748348 0.000162942053437     0 8590.28038257
   7   GALAXY     0.947435678136 0.000110410356425     0 9288.04926471
   8   GALAXY    0.0341969933625 4.01854357817e-05     0  9587.7898065
   9   GALAXY     0.990873544233 0.000111714339476     0 9698.98343697
```

![capture d ecran de 2017-11-27 13-45-16](https://user-images.githubusercontent.com/3166436/33288359-37a4f840-d379-11e7-97f5-990eec897b5b.png)

**Comment by londumas** [comment]
We could use the classification continuum of stars M, K, G, F, A, B, O.
For example if the first best template is A and the second is F or B, then SMALL_DELTA_CHI2=0, else =1

**Comment by londumas** [comment]
If we want we would have to change the following lines:
```
            ii = np.where(tzfit['deltachi2'] < 9)[0]
            tzfit['zwarn'][ii] |= ZW.SMALL_DELTA_CHI2
```

**Comment by sbailey** [comment]
Good catch.  It would probably be sufficient to just not flag SMALL_DELTA_CHI2 between templates that have the same spectype but different subtypes.  I.e. I don't think we need to do the special case bookkeeping of the order of the spectral types and which ones are "between" which others.

This same concept could also apply if we ever split out BGS / ELG / LRG / etc into separate templates with the same spectype but different subtypes.

**Comment by londumas** [comment]
I agree. It is more simple, we still will need to compare a fit to the next from a different class because it also could be a bad spectrum. For exemple in case of Star (chi2=4000), star2 (chi2=4001),qso(chi2=4001) we would like to keep the SMALL_DELTA_CHI2 flag True

**Comment by djschlegel** [comment]
Do you really want to flag even in the case where the class is different?  In BOSS we don't, since there are the AGN (especially at low signal/noise) that fit equally well to a quasar or galaxy.  There are also cases that are equally well fit with a late-type star or nearby galaxy (most of these will be stars, of course).  In BOSS, we made the arbitrary choice of flagging SMALL_DELTA_CHI2 if there was a small chi2 between the best-fit and another fit more than 1000 km/s distant.

**Comment by londumas** [comment]
I looked at the all eBOSS fits:
there are 2,611,704 good objects (zwarn==0)
there are 445,158 not good objects (zwarn!=0)
Among these bad ones 41,072 (i.e. 9.2%) are flagged stars with SMALL_DELTA_CHI2 flag for znum0 and zwarn==0 for znum1, spectype1==spectype2=="star", subtype1!=subtype2

Here is the list of targetid:
[same_star_targetid.txt](https://github.com/desihub/redrock/files/1519091/same_star_targetid.txt)

Here is the distribution of |z-0-z_1|:
![same_star_star_deltaz](https://user-images.githubusercontent.com/3166436/33449000-b165c05e-d5c4-11e7-9fc4-14d86fe7af17.png)

**Comment by sbailey** [comment]
`redrock.fitz` lines 217-220 exclude minima that are within 1000 km/s of another smaller minimum *from the same template*, but we are apparently missing logic for ignoring close minima across different templates when setting `SMALL_DELTA_CHI2`.

**Comment by londumas** [comment]
>  lines 217-220 exclude minima that are within 1000 km/s of another smaller minimum from the same template

Why this condition? Shouldn't we exclude given the SRD?:
`
#   for gal err_syst <0.0002*(1+z)
#   for QSO err_syst <0.0004*(1+z)
`

**Comment by londumas** [comment]
> lines 217-220 exclude minima that are within 1000 km/s of another smaller minimum from the same template

or let me rephrase this: what is the physical motivation for this `1000 km/s`?

**Comment by londumas** [comment]
Among the

**Comment by londumas** [comment]
Among the 41,072, from spplates, we have that their OBJTYPE is:
GALAXY 432
NA 4115
QSO 35632
SKY 72
SPECTROPHOTO_STD 821

**Comment by londumas** [comment]
fixed in PR https://github.com/desihub/redrock/pull/54
